src/hooks/usePendingSubmissions.tsx

```diff
@@ -0,0 +1,263 @@
+    React.useState(0)
+  const [
+    isShowingFailedNotification,
+    showFailedNotification,
+    hideFailedNotification,
+  ] = useBooleanState(false)
+  const [isProcessingPendingQueue, setIsProcessingPendingQueue] =
+    React.useState(false)
+  const [state, setState] = React.useState<{
+    isLoading: boolean
+    loadError: Error | null
+    pendingSubmissions: submissionService.PendingFormSubmission[]
+  }>(defaultState)
+
+  const processPendingQueue = React.useCallback(async () => {
+    if (isMounted.current) {
+      setIsProcessingPendingQueue(true)
+    }
+
+    await submissionService.processPendingQueue({
+      shouldRunExternalIdGeneration: true,
+      shouldRunServerValidation: true,
+    })
+
+    if (isMounted.current) {
+      setIsProcessingPendingQueue(false)
+    }
+  }, [isMounted])
+
+  const reloadPendingSubmissions = React.useCallback(async () => {
+    if (isMounted.current) {
+      setState((currentState) => ({
+        isLoading: true,
+        loadError: null,
+        pendingSubmissions: currentState.pendingSubmissions,
+      }))
+    }
+    let newError = null
+    let newPendingSubmissions: submissionService.PendingFormSubmission[] = []
+
+    try {
+      newPendingSubmissions =
+        await submissionService.getPendingQueueSubmissions()
+    } catch (error) {
+      newError = error as Error
+    }
+
+    if (isMounted.current) {
+      setState({
+        isLoading: false,
+        loadError: newError,
+        pendingSubmissions: newPendingSubmissions,
+      })
+    }
+  }, [isMounted])
+
+  React.useEffect(() => {
+    if (!isPendingQueueEnabled) {
+      return
+    }
+    reloadPendingSubmissions()
+    return submissionService.registerPendingQueueListener(
+      (pendingSubmissions, action) => {
+        switch (action) {
+          case 'SUBMIT_SUCCEEDED': {
+            setSubmittedNotificationCount((current) => current + 1)
+            break
+          }
+          case 'SUBMIT_FAILED': {
+            showFailedNotification()
+            break
+          }
+          case 'SUBMIT_STARTED': {
+            hideFailedNotification()
+            break
+          }
+          case 'ADDITION': {
+            processPendingQueue()
+            break
+          }
+        }
+        setState({
+          isLoading: false,
+          pendingSubmissions,
+          loadError: null,
+        })
+      },
+    )
+  }, [
+    hideFailedNotification,
+    isPendingQueueEnabled,
+    processPendingQueue,
+    reloadPendingSubmissions,
+    showFailedNotification,
+  ])
+
+  React.useEffect(() => {
+    if (!isOffline) {
+      processPendingQueue()
+    }
+  }, [isOffline, processPendingQueue])
+
+  React.useEffect(() => {
+    if (submittedNotificationCount > 0) {
+      const timeoutId = setTimeout(() => {
+        setSubmittedNotificationCount(0)
+      }, 5000)
+      return () => {
+        clearTimeout(timeoutId)
+      }
+    }
+  }, [submittedNotificationCount])
```

Comment:
Could this be done better, ie. without the optimistic timer? (disclaimer: I don't know all the ins and outs)

Reply:
I think ill make the 5 seconds an input prop but default it to 5000. But I would prefer to keep this here as it would be quite annoying to implement with the `submittedNotificationCount` variable being a number.